### DISCUSSION TITLE
FreeBSD support for remote sessions

### DISCUSSION THREAD

**Discussion by noamross** [zed-industries/zed#25601] [Feature Requests]
### Check for existing issues

- [X] Completed

### Describe the feature

This is a request for remote/SSH support for FreeBSD systems.  FreeBSD remote hosts are probably more common than desktops. Currently, attempting to start a remote session with a FreeBSD server results in this error:

```
+ curl -f https://zed.dev/install.sh
+ sh
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  4327  100  4327    0     0  10084      0 --:--:-- --:--:-- --:--:-- 10086
Unsupported platform FreeBSD

⏵ Task `Install zed over ssh` finished with non-zero error code: 1
```

This is more limited than #15309, which is for general support for FreeBSD.

**Comment by attilagyorffy** [comment]
If anyone wants to pick this up, the relevant docs and posts I have found are currently these:
- [SSH Remoting is Here! (Blog)](https://zed.dev/blog/remote-development)
- [Remote Development](https://zed.dev/docs/remote-development#remote-development)

**Comment by ConradIrwin** [comment]
It looks like the remote server doesn't yet compile on FreeBSD, but the first round of compilation errors seem relatively fixable if anyone wants to take this on (the Rust language adapter doesn't know what platform it's running on to download rust analyzer).

Once we have that working we can at least help with the "self-serve" version where you compile and upload the remote server; and I'd be open to figuring out to how to publish freeBSD builds too.

**Comment by attilagyorffy** [comment]
For the time being, I decided to use the [rmate protocol](https://github.com/textmate/rmate) which is quite simple. By running a [local server](https://github.com/zuckschwerdt/zed-rmate-server) and opening files on remote hosts using a [simple shell implementation of the rmate client](https://github.com/aurora/rmate) I can use Zed as my local editor and "push" changes to my remote servers once I hit save. This is quite useful already, even though it doesn't provide a full directory-tree and the capabilities of a full remote session.

**Comment by gplusplus314** [comment]
Hey @ConradIrwin , I got `remote_server` building in FreeBSD. Here are my notes. I also started to start working on the build for all of Zed, but haven't had enough free time. Hope this helps! Please let me know what else I can do to help; I'm a total novice with Rust, but most definitely Rust-curious.

# Building Zed starting from a FreeBSD 14.2 base system.

Assumption is a fresh FreeBSD 14.2 installation with defaults and no extra packages.

## First, as root:
Dependencies for `remote_server`:
```sh
pkg install git rustup-init cmake llvm pkgconf protobuf
```
Dependencies for all of Zed, in addition to above (work in progress; this is incomplete):
```sh
pkg install libx11 alsa-lib
```
## Then, as a standard user:
Cloning into `~/zed`, for example
```sh
rustup-init # pick defaults
cd ~/
git clone https://github.com/zed-industries/zed.git
cd zed
```

Add `"x86_64-unknown-freebsd"` to `rust-toolchain.toml`.

### Build `remote_server`:
```se
cargo build -p remote_server
```

Sanity check:
```sh
cd target/debug
./remote_server --help
```
We then get the following output:
```
Usage: remote_server [COMMAND]

Commands:
  run
  proxy
  version
  help     Print this message or the help of the given subcommand(s)

Options:
  -h, --help  Print help
```

**Reply by notpeter** [reply]
This is amazing.  I would love to see if it's possible to start building FreeBSD x86_64 remote_server binaries as part of our release process via GitHub Actions.

I no experience using actions runners on FreeBSD (or running FreeBSD inside a VM on a self-hosted runner) and would appreciate recommendations / examples of folks doing this to better understand the complexity and current best practices.

**Reply by gplusplus314** [reply]
Official support for FreeBSD with GitHub Actions is still missing. There's a relevant discussion here: https://github.com/actions/runner/issues/385 .

Integrating into the current CI pipeline would obviously be ideal, but I think we (FreeBSD people) can live with an opt-in for providing our own binary. That is, maybe something like `freebsd_remote_server_binary_path` (hopefully a better name than mine) in Zed's settings. Zed could push it up to the target server and run it with the appropriate flags/options.

I put zero effort into naming and structuring the above hypothetical setting, so take it with a grain of salt. But if we're providing our own binaries, we'd also need to specify some more information. Perhaps using a Rust Host Triple might make sense, since those are already well-defined magic strings. Something like this:
```json
{
  "freebsd_binaries": {
    "x86_64-unknown-freebsd": {
      "14.2": "file:///~/path/to/zed/freebsd/amd64/14.2/remote_server",
      "15.0": "https://foobar.com/zed/freebsd/amd64/remote_server"
    }
  }
}
```
Zed could then do the equivalent of a `uname -a` and find the appropriate matching entry in our config.

Or, in the spirit of avoiding complexity, something as simple as this in our settings.json:
```json
{
  "freebsd_remote_server_prompt": true
}
```
... and then Zed would just prompt me for a local path if `remote_server` isn't already on the remote machine; no need for all of the complexity of matching OS version, architecture, etc.

If you steer me a bit, I can try to put some time into this on the weekends. Zero Rust experience here, but I made it through Rustlings. That's all I've got! 😂

**Reply by ConradIrwin** [reply]
Hmm. We do have an environment variable (only in debug builds of zed for now). `ZED_BUILD_REMOTE_SERVER=true` that causes it to use `cross` to compile and build the remote.

We could maybe expose something a little more production-ready through the settings. The main downside is that rust is *slow* to compile things; so what should be a joyous few seconds to connect turns into a frustrating several minutes. It'd be better if we could first check whether the version on the server is up to date, and better still if we were less strict about which versions are compatible.

**Comment by notpeter** [comment]
I decided to take a crack at it.  Please see the [freebsd branch](https://github.com/zed-industries/zed/tree/freebsd) or the PR: 
- https://github.com/zed-industries/zed/pull/29561

I got zed-remote-server to compile in debug mode in ~20mins using [vmactions/freebsd-vm](https://github.com/vmactions/freebsd-vm) on the standard GitHub Actions runners (`ubuntu-latest`) running FreeBSD under QEMU/libvirt.

There's definitely some jank and we may need some (small) changes to upstream.  By default it uses rsync command to copy things into the container, but the command it uses [skips symlinks](https://github.com/notpeter/freebsd-vm/blob/7ed9e507864c057f9317137c7538c96b70728482/run.sh#L235) (and we have symlinks):
```
rsync -avrtopg -e 'ssh -o MACs=umac-64-etm@openssh.com' --exclude _actions --exclude _PipelineMapping  $HOME/work/  $osname:work
```

I'm not super keen on using random github actions and pre-built qemu cow base images from strangers on the internet for building Zed official binaries, but I wanted to see whether it was possible before deciding whether it's practical. Of note, [rust-lang/rustup](https://github.com/rust-lang/rustup/blob/e4f3ad6f893e56ca10f0a329c401a5eafab69607/.github/workflows/ci.yaml#L1334-L1364) is using vmactions/freebsd-vm too, but only for running tests.

Not sure how long a release build would take, and whether whether we can speed them up by using larger runners (`ubuntu-latest` is 4c/16GB) and takes ~20mins.  And this was just `remote_server`, I haven' yet tried building all of zed.  Ideally we'd create a `script/bundle-freebsd` similar to `script/bundle-linux` to better automate the process.

Another idea is to use an alternate CI provider that has native FreeBSD. [Travis-CI](https://docs.travis-ci.com/user/reference/freebsd/) and [Cirrus-CI](https://cirrus-ci.org/guide/FreeBSD/) both advertise FreeBSD support.

**Comment by notpeter** [comment]
Can someone with access to a FreeBSD remote test my builds zed-remote-server builds?

You'll need a supported platform to run the gui (macos, linux), have an ssh accessible remote running freebsd and manually copy a few files.

1. Download [`zed-remote-server-HASH-x86_64-unknown-freebsd.gz.zip`](https://github.com/zed-industries/zed/actions/runs/15069914640/artifacts/3139870184) from [this run](https://github.com/zed-industries/zed/actions/runs/15069914640/job/42367342918). 
2. Unzip, Gunzip, Rename, Copy `zed-remote-server-freebsd-x86_64.gz` 
```
unzip zed-remote-server-*-x86_64-unknown-freebsd.gz.zip
gunzip zed-remote-server-freebsd-x86_64.gz
# rename to match your current zed channel / version
mv zed-remote-server-freebsd-x86_64 zed-remote-server-stable-0.186.9
#mv zed-remote-server-freebsd-x86_64 zed-remote-server-dev-build
ssh user@remote.tld mkdir -p ~/.zed_server
scp zed-remote-server-* user@remote.tld:~/.zed_server/
ssh user@remote.tld 'chmod +x ~/.zed_server/*'
``` 
3. Try connecting with zed `zed ssh://remote.tld/~/somedir/`

Thanks

**Reply by G36maid** [reply]
Yes, It works on my machine!

Arch Linux successfully connected to FreeBSD via Zed 0.186.9.

Screenshot:
![圖片](https://github.com/user-attachments/assets/c5fa7e89-4ed1-43ee-a85f-6a366d24afd1)

**Comment by jbenton** [comment]
Just to confirm, this worked for connecting to my FreeBSD server too!

**Comment by notpeter** [comment]
I merged https://github.com/zed-industries/zed/pull/29561 and cherry-picked it onto Preview.
As of Preview v0.190.1-pre (later today or tomorrow), Zed should support connecting to FreeBSD x86_64 remote hosts over ssh and they should pull down the appropriate pre-built zed-remote-server binaries from our GitHub releases.

**Comment by lakako** [comment]
I have build zed remote v0.197.5 and put it on ~/.zed_server/zed-remote-server-stable-0.197.5

My Client is zed windows v0.197.5

connect command is zed ssh://localfb/~/clash-rsw

but it cause 

<img width="527" height="228" alt="image" src="https://github.com/user-attachments/assets/3f1fce46-d2a0-4c19-8ce8-86939af64a54" />

**Comment by notpeter** [comment]
Sorry, we had some problems where FreeBSD remote server builds were failing so I disabled builds for them here:
- https://github.com/zed-industries/zed/pull/34511

I didn't have a chance to dig in earlier, but I'm wondering whether we double the ram might increase the likelyhood of build success. Trying that here:
- https://github.com/zed-industries/zed/pull/35668

**Reply by notpeter** [reply]
Separately there is also other problems with [vmactions/freebsd-vm](https://github.com/vmactions/freebsd-vm) -- ([example job](https://github.com/zed-industries/zed/actions/runs/16760627724/job/47455781198)).

Where the build gets stuck waiting for a full hour before timing out. Upstream issue:
- https://github.com/vmactions/freebsd-vm/issues/91

If anyone has alternate suggestions for how to build this via CI or would be willing to lend a hand with that freebsd action to make it more reliable, I'd be eternally grateful.

**Reply by notpeter** [reply]
Sadly I couldn't get this to work in CI, it's failing with a cranelift issue.
I can't devote a lot of time to this, but if anyone has ideas or PRs, I'm happy to discuss/review.

**Reply by G36maid** [reply]
It looks like this might be fixed by upstream — see [vmactions/freebsd-vm#91 (comment)](https://github.com/vmactions/freebsd-vm/issues/91#issuecomment-3193547699).
The latest release [v1.2.3](https://github.com/vmactions/freebsd-vm/releases/tag/v1.2.3) includes retry logic for the 200-second timeout, may help.

**Reply by notpeter** [reply]
Yeah, it's still frequently failing with a SIGKILL: [example job](https://github.com/zed-industries/zed/actions/runs/16760627724/job/47513269224).  My assumption is that this is actually because cargo is using too much memory. (or could be something else).

I'm not sure if there's something else we need to do to increase the memory to the freebsd qemu machine or if there are cargo flags we can use to reduce the memory usage / parallelization so it doesn't get killed. 

Separately a better long term solution might be to just cross-compile the freebsd remote using cargo-zigbuild like we do for cross-target zed-remote-server development (e.g. macOS connecting to a linux remote with local developer builds of both).  But that does not yet support FreeBSD as a target:
- https://github.com/rust-cross/cargo-zigbuild/issues/356

**Comment by acama** [comment]
I was able to compile remote_server for `0.207.0` preview on **aarch64** freebsd. I then moved it to `~/.zed_server/zed-remote-server-preview-0.207.0` but when I try to connect I get:
```
Prebuilt remote servers are not yet available for "FreeBSD". See https://zed.dev/docs/remote-development. Please try again.
```
I am on `Zed Preview 0.207.0`

Regarding the `SIGKILL` issue you are getting I think it's very likely because it runs out of RAM, i had the same issue and bumped my qemu VM's ram to 16G (up from 4G).

**Comment by notpeter** [comment]
I'm no longer able to devote time to this, but rust-cross cargo-zigbuild can now cross compile to freebsd so it may be possible to build the freebsd zed-remote-server on macOS/linux hosts. As a result it may be possible to build official FreeBSD zed-remote binaries in CI without resorting to Qemu.

- https://github.com/rust-cross/cargo-zigbuild/issues/356

**Reply by ConradIrwin** [reply]
Ooh! I looked at this nary a fortnight ago and it wasn't ready yet. let me try again

**Comment by JakubKoralewski** [comment]
Building for freebsd aarch64 too would be amazing!

**Comment by xpader** [comment]
I think support remote session for FreeBSD is not too hard for offical.

**Comment by FeironoX5** [comment]
Would be nice to have this!

**Comment by ConradIrwin** [comment]
Someone with a freeBSD box and some time on their hands want to see if the zigbuild version works? If so I can try and integrate that into the release pipeline

**Reply by G36maid** [reply]
I can give the zigbuild version a shot. 
I have a FreeBSD environment ready and can report back if it works(or not) for the remote server.

**Comment by G36maid** [comment]
@ConradIrwin 
The zigbuild version works, but needs a few workarounds. Here's a summary:
Build side (cargo-zigbuild cross-compile from Linux → FreeBSD 15.0):
1. crash-handler/minidumper don't support FreeBSD → gated behind cfg(not(target_os = "freebsd")) with no-op stubs
2. libz-sys always vendors zlib when cross-compiling, but zig cc silently ignores -fPIC for the FreeBSD target, producing non-PIC .o files → patched build.rs to link libz.so from FreeBSD sysroot instead
3. cargo-zigbuild ignores -L paths for -l resolution on FreeBSD → pass full .so paths via RUSTFLAGS -C link-args=
4. Minor cfg gates (gpui queue module) and MaybeUninit fix in fs crate

Client side:

5. parse_platform() in crates/remote/src/transport.rs only recognizes Darwin and Linux — rejects FreeBSD before even checking if a binary exists on the remote. Workaround: override uname on FreeBSD to report Linux x86_64.


Once these are addressed, the remote server runs and SSH remote development works:
<img width="2536" height="1373" alt="2026-04-29-191826_hyprshot" src="https://github.com/user-attachments/assets/fbf904c9-eed5-4dc9-a369-d5373c3b5a0b" />

Full details: I've documented everything in a report — happy to share or open a PR for the  change.
https://gist.github.com/G36maid/c2aff8c1561b307f38f9e1b3aff215e1

**Reply by G36maid** [reply]
Already opened a draft PR #55388
